From 3c3d46adcb62079977ff6c7dfbd345492f2125df Mon Sep 17 00:00:00 2001
From: Eshwar Stalin <estalin@chromium.org>
Date: Tue, 13 Jan 2026 14:22:32 -0800
Subject: [PATCH] [VerticalTabs] Fixing a crash during drag when TabGroup is collapsed

Currently when a tab group is collapsed in Vertical Tabs, we compute the
insertion index incorrectly when the group is collapsed. This leads to a
CHECK failure where we attempt to add a tab inside a group with a
nullopt as group_id. To fix this, we now look at the the position of the
drag and compute the insertion index. This addresses the crash issue we
were encountering.

Bug: 439963720
Change-Id: Ie33bab24616f45e4eef302801c3e0e6502d8b83d
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7464380
Commit-Queue: Eshwar Stalin <estalin@chromium.org>
Reviewed-by: David Pennington <dpenning@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1568708}
---

--- a/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
@@ -168,15 +168,37 @@ void VerticalTabDragHandlerImpl::HandleT
           std::get<const tabs::TabCollection*>(node.GetNodeData()))
           ->GetTabGroup();
   CHECK(tab_group);
-  const int insertion_idx =
+
+  const auto& selection_model = tab_strip_model_->selection_model();
+
+  if (std::all_of(selection_model.selected_tabs().begin(),
+                  selection_model.selected_tabs().end(),
+                  [tab_group](const tabs::TabInterface* selected_tab) {
+                    return selected_tab->GetGroup() == tab_group->id();
+                  })) {
+    // Selected tabs are already in the group, so return early.
+    return;
+  }
+
+  int first_tab_in_group =
       tab_strip_model_->GetIndexOfTab(tab_group->GetFirstTab());
-  const bool should_insert_before_group =
-      tab_strip_model_->IsGroupCollapsed(tab_group->id());
+  int last_tab_in_group =
+      tab_strip_model_->GetIndexOfTab(tab_group->GetLastTab());
+  int first_selected_index =
+      *selection_model.GetListSelectionModel().selected_indices().cbegin();
 
-  tab_strip_model_->MoveSelectedTabsTo(
-      insertion_idx, should_insert_before_group
-                         ? std::nullopt
-                         : std::make_optional(tab_group->id()));
+  if (tab_strip_model_->IsGroupCollapsed(tab_group->id())) {
+    // Selected tabs need to be inserted outside the group if collapsed.
+    int insertion_idx = (first_selected_index < first_tab_in_group)
+                            ? last_tab_in_group
+                            : first_tab_in_group;
+    tab_strip_model_->MoveSelectedTabsTo(insertion_idx, std::nullopt);
+  } else {
+    int insertion_idx = (first_selected_index < first_tab_in_group)
+                            ? first_tab_in_group
+                            : last_tab_in_group;
+    tab_strip_model_->MoveSelectedTabsTo(insertion_idx, tab_group->id());
+  }
 }
 
 void VerticalTabDragHandlerImpl::HandleTabDragOverUnpinnedContainer(
